default/python/AI/pathfinding.py

```diff
@@ -0,0 +1,139 @@
+from heapq import heappush, heappop
+from collections import namedtuple
+
+import freeOrionAIInterface as fo
+from AIDependencies import INVALID_ID
+from turn_state import state
+
+from common.configure_logging import convenience_function_references_for_logger
+(debug, info, warn, error, fatal) = convenience_function_references_for_logger(__name__)
+
+
+# TODO Allow short idling in system to use stationary refuel mechanics in unsupplied systems if we
+#      are very close the target system but can't reach it due to fuel constraints
+# TODO Consider additional optimizations:
+#    - Check if universe.shortestPath complies with fuel mechanics before running
+#      pathfinder (seemingly not worth it - we have exact heuristic in that case)
+#    - Cut off branches that can't reach target due to supply distance
+#    - For large graphs, check existance on a simplified graph (use single node to represent supply clusters)
+#    - For large graphs, check if there are any must-visit nodes (e.g. only possible resupplying system),
+#      then try to find the shortest path between those and start/target.
+def find_path_with_resupply(start, target, fleet_id, minimum_fuel_at_target=0):
+    """Find the shortest possible path between two systems that complies with FreeOrion fuel mechanics.
```

Comment:
Please add comment how enemy fleets and monsters is counted during pathfinding.

Reply:
Currently this doesn't do anything about enemy fleets and monsters (nor does the function this is replacing).  I've been experimenting, though, with some augmentation for this that does take them into account, I plan to post a PR for that shortly after this gets merged.

Reply:
Added a TODO how I have envisioned to handle this - an additional argument for systems that are not to be considered for pathfinding. The logic what those systems are (e.g. those blocked by monsters) is supposed to be figured out by the calling function in advance.